python/cusignal/io/reader.py

```diff
@@ -0,0 +1,328 @@
+# Copyright (c) 2019-2020, NVIDIA CORPORATION.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import cupy as cp
+import numpy as np
+
+import json
+import resource
+
+from mmap import mmap, MAP_PRIVATE, PROT_READ
+
+from ._reader_cuda import _unpack, _pack
+
+
+# https://github.com/cupy/cupy/blob/master/examples/stream/cupy_memcpy.py
+def _pin_memory(array):
+    mem = cp.cuda.alloc_pinned_memory(array.nbytes)
```

Comment:
I'm curious how this differs from what we have in `arraytools` using `numba.cuda.mapped_array`?

Reply:
I'm not sure. I've requested more insight from CuPy developers.
https://github.com/cupy/cupy/issues/3452